File: rust-workspace/libs/corelib/src/domain/term/events.rs
use serde::{Deserialize, Serialize};

use crate::domain::shared::event::EventMeta;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "data")]
pub enum TermDomainEvent {
    TermCreated {
        meta: EventMeta,
        event_name: String,
        taxonomy_id: String,
        parent_id: Option<String>,
        name: String,
        visible: bool,
        description: Option<String>,
    },
    TermUpdated {
        meta: EventMeta,
        event_name: String,
        taxonomy_id: Option<String>,
        parent_id: Option<String>,
        name: Option<String>,
        visible: Option<bool>,
        description: Option<String>,
    },
    TermDeleted {
        meta: EventMeta,
        event_name: String,
    },
    TermSoftDeleted {
        meta: EventMeta,
        event_name: String,
    },
    TermRestoredFromSoftDeleted {
        meta: EventMeta,
        event_name: String,
    },
}

File: node-workspace/libs/corelib/src/domain/identity/events.ts
import { EventMeta } from "../shared/event";


export type IdentityDomainEvent =

    | {
        type: "IdentityCreated";
        data: {
            meta: EventMeta;
            event_name: string;
        };
    }
    | {
        type: "IdentityUpdated";
        data: {
            meta: EventMeta;
            event_name: string;
        };
    }
    | {
        type: "IdentityDeleted";
        data: {
            meta: EventMeta;
            event_name: string;
        };
    }
    | {
        type: "IdentitySoftDeleted";
        data: {
            meta: EventMeta;
            event_name: string;
        };
    }
    | {
        type: "IdentityRestoredFromSoftDeleted";
        data: {
            meta: EventMeta;
            event_name: string;
        };
    };



export const IdentityDomainEventFactory = {
    identityCreated(args: {
        meta: EventMeta;
        event_name: string;
    }): IdentityDomainEvent {
        return { type: "IdentityCreated", data: args };
    },
    identityUpdated(args: {
        meta: EventMeta;
        event_name: string;
    }): IdentityDomainEvent {
        return { type: "IdentityUpdated", data: args };
    },
    identityDeleted(args: {
        meta: EventMeta;
        event_name: string;
    }): IdentityDomainEvent {
        return { type: "IdentityDeleted", data: args };
    },
    identitySoftDeleted(args: {
        meta: EventMeta;
        event_name: string;
    }): IdentityDomainEvent {
        return { type: "IdentitySoftDeleted", data: args };
    },
    identityRestoredFromSoftDeleted(args: {
        meta: EventMeta;
        event_name: string;
    }): IdentityDomainEvent {
        return { type: "IdentityRestoredFromSoftDeleted", data: args };
    },

};